UI/CombatReport/CombatLogWnd.cpp

```diff
@@ -66,52 +66,145 @@ class CombatLogWnd::Impl {
 };
 
 namespace {
-    typedef std::shared_ptr<LinkText> LinkTextPtr;
-
-    const std::string EMPTY_STRING;
-
-    // Counts how many ships or populated planets in \a objects are owned by
-    // each empire. Assumes ownership and populations now are consistent with
-    // whenever the result is supposed to indicate, not accounting for 
-    std::map<int, int> CountByOwner(const std::set<int>& owners, const std::set<int>& objects) {
-        std::map<int, int> objects_per_owner;
-        for (int owner_id : owners)
-            objects_per_owner[owner_id] = 0;
-        for (int obj_id : objects) {
-            auto object = Objects().Object(obj_id); // gets destroyed objects, so this function can be used for initial combat forces even if some are destroyed during the combat
-            if (object && (
-                    object->ObjectType() == OBJ_SHIP || (
-                        object->GetMeter(METER_POPULATION) &&
-                        object->InitialMeterValue(METER_POPULATION) > 0.0f)))
-            {
-                int owner_id = object->Owner();
-                if (!objects_per_owner.count(owner_id))
-                    objects_per_owner[owner_id] = 0;
-                ++objects_per_owner[owner_id];
+
+    // TODO: Function adapted from CombatEvents.cpp, will need to be extracted to a common library
+    const std::string& LinkTag(UniverseObjectType obj_type) {
+        static const std::string EMPTY_STRING("");
+
+        switch (obj_type) {
+        case OBJ_SHIP:
+            return VarText::SHIP_ID_TAG;
+        case OBJ_FLEET:
+            return VarText::FLEET_ID_TAG;
+        case OBJ_PLANET:
+            return VarText::PLANET_ID_TAG;
+        case OBJ_BUILDING:
+            return VarText::BUILDING_ID_TAG;
+        case OBJ_SYSTEM:
+            return VarText::SYSTEM_ID_TAG;
+        case OBJ_FIELD:
+        case OBJ_FIGHTER:
+        default:
+            return EMPTY_STRING;
+        }
+    }
+
+    // TODO: Function adapted from CombatEvents.cpp, will need to beextracted to a common library
+    std::string WrapWithTagAndId(const std::string& meat, const std::string& tag, int id) {
+        return boost::str(boost::format("<%1% %2%>%3%</%1%>") % tag % id % meat);
+    }
+
+    /// Returns true, if the \a object should be counted in initial and/or destroyed empire forces
+    bool ShouldObjectBeShownInEmpireForces(std::shared_ptr<UniverseObject> object) {
+        return object && (object->ObjectType() == OBJ_SHIP ||
+                          (object->GetMeter(METER_POPULATION) &&
```

Comment:
Not sure about the test being if an object has population > 0

More relevant would probably be something similar to the tests used during combat for incapacitation of planets: https://github.com/freeorion/freeorion/blob/master/combat/CombatSystem.cpp#L1161

Reply:
That's the condition that was used before. I just extracted it into a separate function.

Reply:
Actually, the "forces" is a bit ambiguous. Does it aim to enumerate all "active" forces (that is, able to fire), all objects that could be potentially fired upon (all owned planets and ships)?

I fail to see, how I could reuse the code you linked as a general condition over all objects in the system. There's a call to `ObjectCanAttack`, for example. Should I check if any object in the system *can potentially attack* a planet when deciding if that planet should be counted among its owner's forces?

Reply:
The main thing I meant to point to was the tests of various other meters to decide if a planet is incapacitated:
````
                if (target->CurrentMeterValue(METER_SHIELD) <= 0.0f &&
                    target->CurrentMeterValue(METER_DEFENSE) <= 0.0f &&
                    target->CurrentMeterValue(METER_CONSTRUCTION) <= 0.0f)
````
It seemed like the same criteria for deciding that a planet is no longer a valid attacker could be used for marking the planet as "lost" or "disabled" in the summary.

Reply:
The point is, that the label is "Lost" - in case of colonies I would consider only colonies that were completely destroyed (ownership lost) to be "lost".

If you want, I *can* change the forces summary to contain separate sections for "Ships lost" and "Colonies disabled" (or incapacitated).

This will require more changes, including string tables.

-

So far I tried to preserve the object sets reported, just increase verbosity of the report. If you want to change the meaning of the report or the object sets reported - sure, I can do that. I would really like acceptance criteria on the functional side.

I haven't played FO enough to have my own opinion on this topic.

Reply:
I'd also like to point out, that "Lost" objects are populated from `CombatLog::destroyed_object_ids` . I'm not sure it contains (or should contain, for that matter) incapacitated planets.

Reply:
@geoffthemedio , I didn't have the time to revisit this since, but I think I'll refactor this to a more easily extendable form and leave the current (previous) criteria.

I'll make it so that the criteria/groups can be changed in future iterations.